docs/templates/block-template-controller.md

```diff
@@ -0,0 +1,77 @@
+# BlockTemplateController
+[Source file](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/src/BlockTemplatesController.php)
+
+The BlockTemplateController class contains all the business logic which loads the templates into the Site Editor or on the front-end through various hooks available in WordPress & WooCommerce core. Without documenting every method individually, I will look to provide some insight into key functionality. 
+
+## Table of Contents
+
+* [Overview](#overview)
+* [add_block_templates( $query_result, $query, $template_type )](#add_block_templates-query_result-query-template_type-)
+* [get_block_file_template( $template, $id, $template_type )](#get_block_file_template-template-id-template_type-)
+* [render_block_template()](#render_block_template)
+
+## Overview
+
+In the initialization of the class, we hook into the three hooks listed below. These provide us with all of the extensibility points necessary in order to load our own block templates alongside the themes.
+
+Within each method section, I will explain in what scenarios they are executed. 
+
+* filter: `get_block_templates` with `add_block_templates`.
+* filter: `pre_get_block_file_template` with `get_block_file_template`.
+* action: `template_redirect` with `render_block_template`.
+
+## add_block_templates( $query_result, $query, $template_type )
+
+This method is applied to the filter `get_block_templates`, which is executed before returning a unified list of template objects based on a query.
+
+**Typically executed when:**
+* Loading the "All Templates" view in the Site Editor
+* Loading one of the templates on the frontend where the query would build a list of relevant templates based on a hierarchy (for example, the product page hierarchy could be an array containing  `single-product-[product-name].html`, `single-product.html`, `single.html`).
+
+**This method is responsible for:**
+
+* Giving our templates a user-friendly title (e.g. turning "single-product" into "Product Page").
+* It collects all the WooCommerce templates from both the filesystem and the database (customized templates are stored in the database as posts) and adds them to the returned list.
+* In the event the theme has a `archive-product.html` template file, but not category/tag template files, it is eligible to use the `archive-product.html` file in their place. So we trick Gutenberg in thinking some templates (e.g. category/tag) have a theme file available if it is using the `archive-product.html` template, even though _technically_ the theme does not have a specific file for them.
+* If the query is targeting templates for a specific post type (e.g. on the Edit Post view), we do not add ineligible WooCommerce templates to the returned list.
```

Comment:
I'm reading this docs without looking at the source code for added realism 😁 , and I am not sure I understand what this means. Is there any way we can make this clearer?

Reply:
Hmm yes, let me work on rewording this.

Reply:
Hopefully this is an improvement https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5939/commits/c0bf7a0cd668d45986dc11310f6d1f8a3ce1a235